umh-core/pkg/fsm/redpanda/actions.go

```diff
@@ -0,0 +1,303 @@
+
+	// Set the desired state to stopped for the given instance
+	err := b.service.StopRedpanda(ctx)
+	if err != nil {
+		// if the service is not there yet but we attempt to stop it, we need to throw an error
+		return fmt.Errorf("failed to stop Redpanda service %s: %w", b.baseFSMInstance.GetID(), err)
+	}
+
+	b.baseFSMInstance.GetLogger().Debugf("Redpanda service %s stop command executed", b.baseFSMInstance.GetID())
+	return nil
+}
+
+// getServiceStatus gets the status of the Redpanda service
+// its main purpose is to habdle the edge cases where the service is not yet created or not yet running
+func (b *RedpandaInstance) getServiceStatus(ctx context.Context, tick uint64) (redpanda_service.ServiceInfo, error) {
+	info, err := b.service.Status(ctx, tick)
+	if err != nil {
+		// If there's an error getting the service status, we need to distinguish between cases
+
+		if errors.Is(err, redpanda_service.ErrServiceNotExist) {
+			// If the service is being created, we don't want to count this as an error
+			// The instance is likely in Creating or ToBeCreated state, so service doesn't exist yet
+			// This will be handled in the reconcileStateTransition where the service gets created
+			if b.baseFSMInstance.GetCurrentFSMState() == internalfsm.LifecycleStateCreating ||
+				b.baseFSMInstance.GetCurrentFSMState() == internalfsm.LifecycleStateToBeCreated {
+				return redpanda_service.ServiceInfo{}, redpanda_service.ErrServiceNotExist
+			}
+
+			// Log the warning but don't treat it as a fatal error
+			b.baseFSMInstance.GetLogger().Debugf("Service not found, will be created during reconciliation")
+			return redpanda_service.ServiceInfo{}, nil
+		} else if errors.Is(err, redpanda_service.ErrHealthCheckConnectionRefused) {
+			// Instead of conditional state checking, always return a ServiceInfo with failed health checks
+			// This allows the FSM to continue reconciliation and make proper state transition decisions
+			if b.baseFSMInstance.GetCurrentFSMState() != OperationalStateStopped { // no need to spam the logs if the service is already stopped
+				b.baseFSMInstance.GetLogger().Debugf("Health check refused connection for service %s, returning ServiceInfo with failed health checks", b.baseFSMInstance.GetID())
+			}
+			infoWithFailedHealthChecks := info
+			infoWithFailedHealthChecks.RedpandaStatus.HealthCheck.IsLive = false
+			infoWithFailedHealthChecks.RedpandaStatus.HealthCheck.IsReady = false
+			// Return ServiceInfo with health checks failed but preserve S6FSMState if available
+			return infoWithFailedHealthChecks, nil
+		}
+
+		// For other errors, log them and return
+		b.baseFSMInstance.GetLogger().Errorf("error updating observed state for %s: %s", b.baseFSMInstance.GetID(), err)
+		return redpanda_service.ServiceInfo{}, err
+	}
+
+	return info, nil
+}
+
+// updateObservedState updates the observed state of the service
+func (b *RedpandaInstance) updateObservedState(ctx context.Context, tick uint64) error {
```

Comment:
here in this function there will be changes in benthos and s6 as well, as for some reasons this is sometimes having large delays taking down the entire reconcile loop (see also slack)